source/common/common/inline_map.h

```diff
@@ -0,0 +1,460 @@
+                                          InlineMap(registry));
+  }
+
+private:
+  friend class InlineMapDescriptor<StorageKey>;
+
+  InlineMap(TypedInlineMapDescriptor& registry) : descriptor_(registry) {
+    // Initialize the inline entries to nullptr.
+    uint64_t inline_keys_num = descriptor_.inlineKeysNum();
+    inline_entries_valid_.resize(inline_keys_num, false);
+    memset(inline_entries_storage_, 0, inline_keys_num * sizeof(StorageValue));
+    inline_entries_ = reinterpret_cast<StorageValue*>(inline_entries_storage_);
+  }
+
+  template <class ArgV> void resetInlineMapEntry(uint64_t inline_entry_id, ArgV&& value) {
+    ASSERT(inline_entry_id < descriptor_.inlineKeysNum());
+
+    clearInlineMapEntry(inline_entry_id);
+
+    // Construct the new entry in the inline array.
+    new (inline_entries_ + inline_entry_id) StorageValue(std::forward<ArgV>(value));
+    setInlineEntryValid(inline_entry_id, true);
+  }
+
+  uint64_t clearInlineMapEntry(uint64_t inline_entry_id) {
+    ASSERT(inline_entry_id < descriptor_.inlineKeysNum());
+
+    // Destruct the old entry if it is valid.
+    if (inlineEntryValid(inline_entry_id)) {
+      (inline_entries_ + inline_entry_id)->~StorageValue();
+      setInlineEntryValid(inline_entry_id, false);
+      return 1;
+    }
+
+    return 0;
+  }
+
+  template <class ArgK> absl::optional<uint64_t> inlineLookup(const ArgK& key) const {
+    const auto& map_ref = descriptor_.inlineKeysMap();
+    if (auto iter = map_ref.find(key); iter != map_ref.end()) {
+      return iter->second.inlineId();
+    }
+    return absl::nullopt;
+  }
+
+  bool inlineEntryValid(uint64_t inline_entry_id) const {
+    ASSERT(inline_entry_id < descriptor_.inlineKeysNum());
+    return inline_entries_valid_[inline_entry_id];
+  }
+  void setInlineEntryValid(uint64_t inline_entry_id, bool flag) {
+    ASSERT(inline_entry_id < descriptor_.inlineKeysNum());
+
+    if (flag) {
+      inline_entries_size_++;
+    } else {
+      ASSERT(inline_entries_size_ > 0);
+      inline_entries_size_--;
+    }
+
+    inline_entries_valid_[inline_entry_id] = flag;
+  }
+
+  // This is the reference of the registry that the inline map created by. This is used to
+  // validate the inline handle validity and get the inline key set.
+  const TypedInlineMapDescriptor& descriptor_;
+
+  // This is the underlay hash map for the dynamic map entries.
+  DynamicHashMap dynamic_entries_;
+
+  // This is flags to indicate if the inline entries are valid.
+  // TODO(wbpcode): this will add additional one time memory allocation when constructing inline
+  // map. It is possible to use memory in the inline_entries_storage_ to store the flags in the
+  // future. But it should be fine for now.
+  std::vector<bool> inline_entries_valid_;
+
+  uint64_t inline_entries_size_{};
+
+  StorageValue* inline_entries_{};
+
+  // This should be the last member of the class and no member should be added after this.
+  uint8_t inline_entries_storage_[];
+};
```

Comment:
I'm guessing this makes it so you can't copy-constructor copy-assign a map, but can you check that?

I'm not sure about moving a map. Can you check that too? Of course you can move an InlineMapPtr.

Speaking of which, should we add an API to make a copy of a map?

Reply:
I think the `InlineStorage` has forbidden any copy. And all map objects should be managed by the unique ptr (map object is only created by `create` method which will return a unique ptr of map), so we should never move a map directly.

Reply:
oh good point -- you inherit that property from InlineStorage. But WDYT about adding an explicit copy method that returns a new unique_ptr?

Reply:
I am OK to it. But I didn't get in which case we will try to copy a inline map instance. I am inclined to leave this to the future, when we accually need it.